modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxSavepointsTransactionalCacheTest.java

```diff
@@ -0,0 +1,562 @@
+
+/** */
+public abstract class TxSavepointsTransactionalCacheTest extends GridCacheAbstractSelfTest {
+    /** {@inheritDoc} */
+    @Override protected int gridCount() {
+        return 3;
+    }
+
+    /** */
+    protected CacheConfiguration<Integer, Integer> getConfig() {
+        CacheConfiguration<Integer, Integer> cfg = new CacheConfiguration<>();
+
+        cfg.setAtomicityMode(TRANSACTIONAL);
+
+        cfg.setCacheMode(cacheMode());
+
+        cfg.setName(cacheMode().name());
+
+        cfg.setBackups(2);
+
+        cfg.setWriteSynchronizationMode(CacheWriteSynchronizationMode.FULL_SYNC);
+
+        return cfg;
+    }
+
+    /** Override this method to use different cache modes for tests.*/
+    protected abstract CacheMode cacheMode();
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testGet() throws Exception {
+        final IgniteCache<Integer, Integer> cache = grid(2).getOrCreateCache(getConfig());
+
+        for (TransactionConcurrency concurrency : TransactionConcurrency.values())
+            for (TransactionIsolation isolation : TransactionIsolation.values()) {
+                try (Transaction tx = grid(2).transactions().txStart(concurrency, isolation)) {
+                    info(concurrency + " " + isolation);
+
+                    tx.savepoint("sp");
+
+                    assertEquals("Broken savepoint in " + concurrency + " " + isolation +
+                        " transaction.", null, cache.get(1));
+
+                    tx.rollbackToSavepoint("sp");
+
+                    IgniteInternalFuture<Boolean> fut = GridTestUtils.runAsync(new Callable<Boolean>() {
+                        @Override public Boolean call() throws Exception {
+                            return cache.putIfAbsentAsync(1, 1).get();
+                        }
+                    },cacheMode().name()+"_get");
+
+                    assertTrue(fut.get(3_000));
+
+                    assertEquals("Broken multithreaded rollback to savepoint in " + concurrency +
+                        " " + isolation + " transaction.", (Integer) 1, cache.get(1));
+
+                    tx.commit();
+                } finally {
+                    cache.remove(1);
+                }
+            }
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testPut() throws Exception {
+        final IgniteCache<Integer, Integer> cache = grid(2).getOrCreateCache(getConfig());
+
+        GridTestSafeThreadFactory factory = new GridTestSafeThreadFactory(cacheMode().name()+"_put");
+
+        for (TransactionConcurrency concurrency : TransactionConcurrency.values())
+            for (TransactionIsolation isolation : TransactionIsolation.values()) {
+                try (Transaction tx = grid(2).transactions().txStart(concurrency, isolation)) {
+                    info(concurrency + " " + isolation);
+
+                    tx.savepoint("sp");
+
+                    assertTrue(cache.putIfAbsent(1, 0));
```

Comment:
every time you start a test, cache must be empty. It is guaranteed in beforeTest().
So you don't have to check it in your test